### PR TITLE
FM-254: Create checkpoint

### LINKED ISSUE #254: ExecInterpreter::end construct a BottomUpCheckpoint and add it to the ledger
As a subnet operator I want the Fendermint interpreter to create a `BottomUpCheckpoint` at the end of the checkpointing period and pass it to #248

### PR DESCRIPTION
**Pull request by aakoshh**
Closes #254 
Depends on https://github.com/consensus-shipyard/ipc-solidity-actors/pull/202

Changes:
- [x] Check if the gateway is installed
- [x] Detect checkpoint period boundary
- [x] Fetch the current validator set
- [x] Create a Merkle tree out of the validator set
- [x] Construct checkpoint
- [x] Insert checkpoint into the ledger

I created a support data structure called `ValidatorMerkleTree` using the library we found. Had to [fork it](https://github.com/consensus-shipyard/merkle-tree-rs) to update it to the latest ethers. There are many things we could change in that library to avoid the multitude of panic conditions. It only works for the happy path, but for now that is all we need. It's also a bit annoying that it converts resulting bytes into strings that I have to parse back bytes, but we can deal with these later.

ABI bindings built with:
```bash
export IPC_ACTORS_TAG=origin/fendermint247
make clean-ipc-actors-abi ipc-actors-abi
```

### CODE REVIEW

**Review by dnkolegov — APPROVED** [review]

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    match should_create_checkpoint(gateway, state, height)? {
+        None => Ok(None),
+        Some(subnet_id) => {
+            // Get the current power table.
+            let power_table = power_table(client, height)
+                .await
+                .context("failed to get the power table")?;
+
+            // TODO #252: Take the next changes from the gateway.
+            let power_updates = PowerUpdates(Vec::new());
```
Comment (adlrocha) [inline fendermint/vm/interpreter/src/fvm/checkpoint.rs:58]: We assume that the updates have already been submitted and queued in the child's gateway through the top-down proof of finality, right?

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            // TODO: #252: Take the configuration number of the last change.
+            let next_configuration_number = 0;
+
+            // Construct checkpoint.
+            let checkpoint = BottomUpCheckpoint {
+                subnet_id,
+                block_height: height.value(),
+                block_hash,
+                next_configuration_number,
+                cross_messages_hash: et::H256::zero().0,
```
[the remark this answers is not in the document]
Comment (aakoshh) [reply fendermint/vm/interpreter/src/fvm/checkpoint.rs:69]: Thanks for pointing this out. I was hoping that you guys can tell me in the review where to get this from. Is there a method to call to retrieve all the messages?

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs (hunk 1 above)
Comment (aakoshh) [reply fendermint/vm/interpreter/src/fvm/checkpoint.rs:58]: Yes, this takes the updates placed in the gateway by the top-down syncer.

File: fendermint/vm/interpreter/src/fvm/checkpoint.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (aakoshh) [reply fendermint/vm/interpreter/src/fvm/checkpoint.rs:69]: I created https://github.com/consensus-shipyard/fendermint/issues/281 to assign a value to this field. 

@adlrocha yes, the relayer is expected to retrieve the cross-messages and send them with the checkpoint in https://github.com/consensus-shipyard/fendermint/issues/257

The parent gateway is expected to hash the messages and compare them to the inside the checkpoint.

**Pull request merged**